draft-thomson-http-oblivious.md

```diff
@@ -159,10 +159,110 @@ occur, as shown in {{fig-overview}}:
 
 # HPKE Encapsulation
 
+HTTP message encapsulation uses HPKE for request and response encryption.
+An encapsulated HTTP message includes the following values:
+
+1. A binary-encoded HTTP message [CITEME].
+2. Padding of arbitrary length which MUST contain all zeroes.
+
+The encoding of an HTTP message is as follows:
+
+~~~
+Plaintext Message {
+  Message Length (i),
+  Message (..),
+  Padding Length (i),
+  Padding (..),
+}
+~~~
+
+This structure is then encrypted under a key with a specific identity, forming an encapsulated HTTP
+message, with the following structure:
+
+~~~
+Key Identifer {
+  Key ID Length (i),
+  Key ID (..),
+}
+
+Encapsulated Message {
+  Key Identifier (..)
+  Encrypted Message Length (i),
+  Encrypted Message (..),
+}
+~~~
+
+The encapsulated message Key ID, as well as the encryption mechanics, are different for requests
+and responses, as described below.
+
 ## HPKE Encapsulation of Requests {#request}
 
+Clients encapsulate a request Plaintext Message `msg` with an HPKE public key `pkR`, whose Key Identifier
+is `keyID` as follows:
+
+1. Compute an HPKE context using `pkR`, yielding `context` and encapsulation key `enc`
+2. Encrypt (seal) `msg` with `keyID` as associated data using `context`, yielding ciphertext `ct`
+3. Concatenate `enc` and `ct`, yielding an Encrypted Message `encrypted_msg`
+
+In pseudocode, this procedure is as follows:
+
+~~~
+enc, context = SetupBaseS(pkR, "request")
+aad = 0x01 || keyID
+encrypted_msg = context.Seal(aad, msg)
+~~~
+
+Clients construct the Encapsulated Message `req` using `keyID` and `encrypted_msg`.
```

Comment:
Yeah, I was going for brevity in the textual descriptions, but we make it more complete to match the pseudocode. 

> Reading the -07 HPKE draft, it doesn't include enc in the output of context.Seal. Should it?

I don't think so, since you might want to use the same context for multiple seals but don't want to send `enc` alongside each ciphertext.